File: mcp-bridge.js
const { spawn } = require('child_process');

// 1. Configure Yunxiao access parameters (prefer env variables)
const YUNXIAO_TOKEN = process.env.YUNXIAO_ACCESS_TOKEN || "YOUR_YUNXIAO_ACCESS_TOKEN";
const ORG_ID = process.env.YUNXIAO_ORG_ID || "YOUR_ORGANIZATION_ID";
const PROJECT_ID = process.env.YUNXIAO_PROJECT_ID || "YOUR_PROJECT_ID";

// 2. Spawn the MCP Server process
const child = spawn('npx', ['-y', 'alibabacloud-devops-mcp-server'], {
  env: { ...process.env, YUNXIAO_ACCESS_TOKEN: YUNXIAO_TOKEN },
  shell: true
});

let stdoutData = '';

function sendJsonRpc(msg) {
  child.stdin.write(JSON.stringify(msg) + '\n');
}

// 3. Handle incoming JSON-RPC responses from stdout
child.stdout.on('data', (data) => {
  stdoutData += data.toString();
  const lines = stdoutData.split('\n');
  stdoutData = lines.pop(); // Keep the incomplete line
  
  for (const line of lines) {
    if (!line.trim()) continue;
    try {
      const msg = JSON.parse(line);
      
      // Step A: When initialized, send the initialized notification and invoke the target tool
      if (msg.id === 1) {
        sendJsonRpc({ jsonrpc: "2.0", method: "notifications/initialized" });
        
        // Example: Search work items
        sendJsonRpc({
          jsonrpc: "2.0",
          id: 2,
          method: "tools/call",
          params: {
            name: "search_workitems",
            arguments: {
              organizationId: ORG_ID,
              category: "Req,Task,Bug",
              spaceId: PROJECT_ID,
              perPage: 100
            }
          }
        });
      } 
      // Step B: Receive tool execution results and output
      else if (msg.id === 2) {
        if (msg.error) {
          console.error("MCP Tool Call Error:", JSON.stringify(msg.error, null, 2));
          process.exit(1);
        } else {
          // Output the raw text result for the calling agent to parse
          console.log(msg.result.content[0].text);
        }
        child.kill();
        process.exit(0);
      }
    } catch (err) {
      // Ignore non-JSON logs (e.g. npx installation downloads or status info)
    }
  }
});

// 4. Initiate the standard MCP connection handshake
sendJsonRpc({
  jsonrpc: "2.0",
  id: 1,
  method: "initialize",
  params: {
    protocolVersion: "2024-11-05",
    capabilities: {},
    clientInfo: { name: "agent-bridge-client", version: "1.0.0" }
  }
});

// 5. Timeout safeguard to terminate the subprocess
setTimeout(() => {
  child.kill();
  process.exit(1);
}, 20000);

File: mcp-bridge.py
import subprocess
import json
import os
import sys

# 1. Configure Yunxiao access parameters (prefer env variables)
YUNXIAO_TOKEN = os.environ.get("YUNXIAO_ACCESS_TOKEN", "YOUR_YUNXIAO_ACCESS_TOKEN")
ORG_ID = os.environ.get("YUNXIAO_ORG_ID", "YOUR_ORGANIZATION_ID")
PROJECT_ID = os.environ.get("YUNXIAO_PROJECT_ID", "YOUR_PROJECT_ID")

# 2. Spawn the MCP Server process (using npx)
env = os.environ.copy()
env["YUNXIAO_ACCESS_TOKEN"] = YUNXIAO_TOKEN

# Note: shell=True is required on Windows to look up npx properly
kwargs = {
    "stdin": subprocess.PIPE,
    "stdout": subprocess.PIPE,
    "stderr": subprocess.PIPE,
    "text": True,
    "env": env,
    "shell": True
}
if os.name == 'nt':
    kwargs["creationflags"] = subprocess.CREATE_NO_WINDOW

process = subprocess.Popen(
    ["npx", "-y", "alibabacloud-devops-mcp-server"],
    **kwargs
)

def send_msg(msg):
    process.stdin.write(json.dumps(msg) + "\n")
    process.stdin.flush()

# Step 1: Send the standard MCP initialize request
send_msg({
    "jsonrpc": "2.0",
    "id": 1,
    "method": "initialize",
    "params": {
        "protocolVersion": "2024-11-05",
        "capabilities": {},
        "clientInfo": {"name": "python-bridge", "version": "1.0.0"}
    }
})

# Step 2: Listen on stdout line-by-line and process JSON-RPC
for line in iter(process.stdout.readline, ""):
    if not line.strip():
        continue
    try:
        msg = json.loads(line)
        if msg.get("id") == 1:
            # Server successfully initialized. Send notification and call the tool.
            send_msg({"jsonrpc": "2.0", "method": "notifications/initialized"})
            send_msg({
                "jsonrpc": "2.0",
                "id": 2,
                "method": "tools/call",
                "params": {
                    "name": "search_workitems",
                    "arguments": {
                        "organizationId": ORG_ID,
                        "category": "Req,Task,Bug",
                        "spaceId": PROJECT_ID,
                        "perPage": 100
                    }
                }
            })
        elif msg.get("id") == 2:
            # Process tool execution results
            if "error" in msg:
                print(f"Error: {msg['error']}", file=sys.stderr)
                sys.exit(1)
            else:
                print(msg["result"]["content"][0]["text"])
            break
    except json.JSONDecodeError:
        pass

process.terminate()
